render_dsdl.py

```diff
@@ -0,0 +1,273 @@
+matching = list(filter(lambda t: fnmatch(t.full_name, pattern), parsed_types))
+if not matching:
+    die('No types match the pattern: %s' % pattern)
+
+# Natural sorting by name and version (numerical ordering, not lexicographical);
+# newest version first, oldest version last.
+matching.sort(key=lambda t: ([int(c) if c.isdigit() else c for c in re.split(r'(\d+)', t.full_name)] +
+                             [-t.version.major, -t.version.minor]))
+
+# See if we were asked to render a particular type only.
+# If that is the case, output an abridged form, provide a reference, and then exit.
+# The output should contain the latest non-deprecated definition.
+if '*' not in pattern:
+    matching_except_deprecated = list(filter(lambda t: not t.deprecated, matching))
+    if not matching_except_deprecated:
+        die('All versions of the type %s are deprecated, nothing to display' % pattern)
+
+    t = matching_except_deprecated[0]       # Due to sorting, newest version ends up first
+    service_or_subject = 'service' if isinstance(t, pydsdl.data_type.ServiceType) else 'subject'
+    print(r'The DSDL source text of \verb|%s|' % t.full_name)
+    print('version %d.%d' % t.version)
+    if len(matching) > 2:
+        print('(there are', len(matching) - 1, 'older versions)')
+    elif len(matching) > 1:
+        print('(there is one older version)')
+    else:
+        print('(this is the only version)')
+
+    if t.has_fixed_port_id:
+        print('with a fixed', service_or_subject, 'ID', t.fixed_port_id)
+    else:
+        print('without fixed', service_or_subject, 'ID')
+
+    print(r'is provided below.')
+    print(r'More information is available in')
+    print(r'section \ref{sec:dsdl:%s} on page \pageref{sec:dsdl:%s}.' % (t.full_name, t.full_name))
+    print(r'\pagebreak[2]{}')           # This is needed to discourage page breaks within the listings
+    print(render_dsdl_definition(t))
+    exit(0)
+
+# Group by namespace and by type name (i.e., all versions grouped together by name).
+# We re-sort (remember, the sorting is stable) to move types that have at least one version with a fixed port ID
+# to the top.
+grouped = OrderedDict()
+for t in sorted(matching, key=lambda t: not t.has_fixed_port_id):
+    grouped.setdefault(t.namespace, OrderedDict()).setdefault(t.full_name, []).append(t)
+
+# Render short reference
+naked_pattern = pattern.strip('.*')
+is_nested_namespace = '.' in naked_pattern
+
+# We avoid using longtabu unless we really have to because it's a bloody disaster. Its caption is always misaligned
+# vertically (requires hacking with \abovecaptionskip to fix) and it tends to run over text and footnotes below it.
+table_environment = 'tabu' if is_nested_namespace else 'longtabu'
+
+if is_nested_namespace:
+    # Confine nested namespace indexes to one page, assuming that they are always compact enough to fit.
+    print(r'{\parindent=-\leftskip\begin{minipage}{\textwidth}\centering')
+
+print(r'\begin{ThreePartTable}')
```

Comment:
Sure. I would avoid jinja though because it's an external dependency, and instead used something like [Pyratemp](https://www.simple-is-better.org/template/pyratemp.html) because it's MIT-licensed and distributed as a one-file module which can be re-distributed together with the script (it's used in libuavcan and libcanard for code generation). That said, I don't want to change the script now.